runtime/near-vm-logic/src/vmstate.rs

```diff
@@ -181,3 +181,116 @@ impl Registers {
         Ok(())
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::Registers;
+
+    use crate::gas_counter::GasCounter;
+
+    use near_primitives_core::config::{ExtCostsConfig, VMLimitConfig};
+    use near_vm_errors::HostError;
+
+    struct RegistersTestContext {
+        gas: GasCounter,
+        cfg: VMLimitConfig,
+        regs: Registers,
+    }
+
+    impl RegistersTestContext {
+        fn new() -> Self {
+            let costs = ExtCostsConfig::test();
+            Self {
+                gas: GasCounter::new(costs, u64::MAX, 0, u64::MAX, false),
+                cfg: VMLimitConfig::test(),
+                regs: Default::default(),
+            }
+        }
+
+        #[track_caller]
+        fn set(&mut self, register_id: u64, value: &str) {
+            self.regs.set(&mut self.gas, &self.cfg, register_id, value.as_bytes()).unwrap();
+            self.read(register_id, Some(value));
+        }
+
+        #[track_caller]
+        fn set_failure(&mut self, register_id: u64, value: &str) {
+            let want = Err(HostError::MemoryAccessViolation.into());
+            let got = self.regs.set(&mut self.gas, &self.cfg, register_id, value.as_bytes());
+            assert_eq!(want, got);
+        }
+
+        #[track_caller]
+        fn read(&mut self, register_id: u64, value: Option<&str>) {
```

Comment:
I think this code would become slightly more readable if the methods largely concerned with asserting contained `assert_` or `check_` in them. That is `assert_read` or `check_read`, `assert_set_fails`, `assert_gas`.

Not a huge deal, though, since this is pretty self-contained.